### ISSUE TITLE
allow using categorical data for color, marker, size, …

### ISSUE THREAD

**Issue by flying-sheep** [matplotlib/matplotlib#6214]
currently specifying an array for color or size, only works for continuous data. using markers and colors mapped to subsets based on categorical data is hard and manual (basically you subset the data in a loop and call plot multiple times).

i thought it would be hard as list-of-string already means something for some parameters (markers and colors can be specified as strings). but there is pandas’ [category dtype](http://pandas.pydata.org/pandas-docs/stable/categorical.html).

i propose that using a categorical vector for one of these parameters should be automatically mapped to a palette of colors/markers when specified as `c` or `marker`.

```py
# with A, B, C, being either categorical or quantitative,
# S being quantitative, and M being categorical
plt.scatter('A', 'B', s='S', c='C', marker='M', data=df)

# of course also works with arrays:
plt.scatter(arr[:,0], arr[:,1], c=arr[:,-1] > 0)
```

the color palette for categorical data would be `plt.rcParams['axes.prop_cycle'].by_key()['color']`, the marker palette a new `rcParam`:

**Comment by flying-sheep** [comment]
@WeatherGod said:

> I am not sure what you mean by "this ship has sailed". All we implemented was a generic way to *access* data within a dataframe or other similar objects, and a minimal usage. I don't see any limitations to extending the syntax, but that'll have to be done judiciously, like how this first round of the API was done. There are a few technical improvements that will be needed (having multiple norms, be able to modify more properties, etc), but we have been wanting to implement them for awhile now.

**Comment by flying-sheep** [comment]
well, see the open questions and the fact that allowing to use `scatter(marker='ColName', data=...)` is a breaking change since the column could be named `'o'`…

**Comment by WeatherGod** [comment]
Yes, in the face of ambiguity, we resist the temptation to guess. However,
that doesn't mean that the problem can't be solved somehow (flags? other
annotated types? etc.) We are open to solutions to these very real
problems. We just don't want to slap together a haphazard solution that we
would be stuck with and regret later.

We just need someone to have an inspired idea on how to solve it.

**Comment by tacaswell** [comment]
We also have a solicitation up for a GSoC student to work on categorical data.

Plotting methods in mpl should never create a color bar or legend, those are (by design) separate steps.

You can pass in a `BoundryNorm` for the color mapping which will give discrete colors (and the color bar code will show them as solid chunks).

Looping over a groupby does not seem that onerous.  It also gives you the flexbility to do _any_ type of plotting in your loop (not just `scatter`, ex `quiver`, `bar`, etc).  Thus instead of putting the logic in _every_ plotting function, the user gets to own the complexity of their data reduction (which will _always_ boil down to a numpy array) and then mpl will plot it.

The core issue is that if you have a very restricted sort of data (tabular, many columns, lots of categorical) it may make sense to push many of the groupby and aggregation operations down into the plotting.  This is my understanding of what people like about ggplot and is definitely the case with `vegalite`.  When working on https://github.com/ellisonbg/altair the hardest part of translating their spec to be implemented on top of mpl was not the plotting, but sorting out the aggregations/computation aspect of it.  Those complexities are what I think MPL should not pick up.  One of the reasons for the wide ranging success of mpl is precisely that it stays _out_ (to a large degree, hist and contour are major exceptions) of computation.

As I said in the other thread we are very positive on domain specific charting libraries to be built on top of mpl, but the core team does not have the bandwidth (or domain knowledge in many cases) to write them.

**Comment by flying-sheep** [comment]
> Thus instead of putting the logic in every plotting function

refactoring the plotting functions to work more grammar-of-graphics-like would manage to do that as well.

1. raw data and parameters as given to function
2. mapping of dataframe-style representation of relevant data data to

	* one or more geometries (bar, scatter, image, …)
	* a set of aesthetics per geometry (colors, markers, size, …)

3. representation of graphics

currently i guess they skip step 2 and go directly from 1 to 3, leading to duplicate effort, am i right?

**Comment by story645** [comment]
@flying-sheep would a colormap dict work. I'm thinking of something in the style of [ListedColormap](http://matplotlib.org/api/colors_api.html#matplotlib.colors.ListedColormap) ala:

```
cmap, norm = mcolors.DictColor({'snow':'palegray', 'rain':'navyblue, 'broken detector':red})
```

that can then be hooked into colormaps (and possibly legends...). My main qualm with it is that it would have to do both cmap and norm simultaneously, but mcolors handles both the cmap & norm code and everything that takes a cmap arg takes a norm arg, so that may not be much overhead.

**Comment by flying-sheep** [comment]
sounds good!

**Comment by naught101** [comment]
At the moment, I can run `scatter([0,1,2], [0,1,2], c=[0,1,2])`, which results in a nice plot, and I can run `scatter([0,1,2], ['A','B','C'], c=[0,1,2])`, which also results in a nice plot, with categories atomatically mapped to the y-axis. 

`scatter([0,1,2], [0,1,2], c=['A','B','C'])` doesn't work though. It would be *really* nice if it did, with each category automatically mapped to a colour. That would make things like these questions much easier to answer: https://stackoverflow.com/questions/26139423/plot-different-color-for-different-categorical-levels-using-matplotlib, https://stackoverflow.com/questions/31883097/elegant-way-to-match-a-string-to-a-random-color-matplotlib

Not sure if that should be submitted as a separate issue or not.

**Comment by bla1089** [comment]
Why not just use something like this?

cdict = {k: v for k, v in enumerate(np.unique([<categories>])}

cvals = [cdict[c] for c in categories]

scatter ([0, 1, 2], [0, 1, 2], c=cvals)


I agree that it would be a nice convenience. I also suspect that there will
be a lot of type and value checking that must go into an automated routine.

Ben

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by story645** [comment]
@naught101 the thing that makes it difficult is the reason it's not implemented, namely that it would require extending unit support to all scalermappable objects (so functions like imshow and pcolor and anything else with a colorbar) and then hooking in a way to generate legends off of colormaps. It's definitely on my wishlist for categoricals, but on the backburner while we refine the current implementation.

**Comment by naught101** [comment]
@bla1089: because then the legend doesn't have automatic labels.

**Comment by bla1089** [comment]
@naught101 try something like

reverse_cdict = {v: k for k, v in cdict.iteritems()}

handles, labels = ax.get_legend_handles_labels()

labels = [reverse_cdict[x] for x in labels]

ax.legend(handles, labels)


Ben

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by naught101** [comment]
@bla1089: Yeah, that's what I'm doing with a custom function. It would be good enough in a lot of cases to have a simple function that mapped categories to colours and returned `(cdict, labels)` or something like that..

**Comment by timhoffm** [comment]
While we’re trying to extend our API with reasonable higher-level extensions (e.g. categorical data), at its core matplotlib is a low level plotting library. For the time being, you might want to have a look at [seaborn](http://seaborn.pydata.org/) for more specialized data science related plots.

**Comment by mwaskom** [comment]
>  For the time being, you might want to have a look at seaborn for more specialized data science related plots.

In fact this was recently added to seaborn and will be available in the next release: https://github.com/mwaskom/seaborn/pull/1436

**Comment by dburkhardt** [comment]
I this this functionality has been added at some point and should be closed.

The following code works as intended:

```python
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

data = np.random.normal(size=(100,2))
colors = pd.Series(data[:,0], dtype='category')
sizes = pd.Series(data[:,1], dtype='category')

plt.scatter(data[:,0], data[:,1], c=colors, s=sizes)
```

**Comment by story645** [comment]
@dburkhardt that data is a float, which should have always worked barring some bugs on pandas series unpacking.  Here categorical is used interchangeably with string data types and string datatypes still don't work.

**Comment by dburkhardt** [comment]
I see, this issue is essentially asking for a matplotlib implementation of MATLAB's gscatter: https://www.mathworks.com/help/stats/gscatter.html. So this desired functionality actually goes back >10 years, I think (http://matplotlib.1069221.n5.nabble.com/matplotlib-gscatter-td12596.html).

**Comment by jklymak** [comment]
... note that Matlab put `gscatter` into a domain-specific toolbox.

**Comment by dburkhardt** [comment]
Good point @jklymak, but that toolbox, Statistics and Machine Learning, is also where Mathworks put boxplots (https://www.mathworks.com/help/stats/boxplot.html) which are provided in matplotlib (https://matplotlib.org/api/_as_gen/matplotlib.pyplot.boxplot.html).

Either way, I also would make use of this functionality if it did make it's way into matplotlib, but unfortunately my programming skills are not strong enough for me to be confident in implementing it so I can't do much more than +1 the topic.

**Comment by jklymak** [comment]
There is a cost-benefit to this sort of thing.  It looks like seaborn has this, which is a more stats-oriented package than matplotlib, so having them maintain it is great.  If I were doing this, I'd just do something like the following, which is hardly more work than the proposed interface:

```python

categories =['A', 'B', 'C']
colors=['r', 'g', 'b']
symb = ['d', 's', 'o']

for ind, cat in enumerate(categories):
    in = np.where(data.cat==cat)
   ax.scatter(data.x[in], data.y[in], c=colors[ind], markerstyle=symb[ind])

```
